core/src/main/java/org/geysermc/geyser/registry/populator/CustomItemRegistryPopulator.java

```diff
@@ -0,0 +1,328 @@
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.registry.populator;
+
+import com.nukkitx.nbt.NbtMap;
+import com.nukkitx.nbt.NbtMapBuilder;
+import com.nukkitx.nbt.NbtType;
+import com.nukkitx.protocol.bedrock.data.inventory.ComponentItemData;
+import com.nukkitx.protocol.bedrock.data.inventory.ItemData;
+import com.nukkitx.protocol.bedrock.packet.StartGamePacket;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import org.geysermc.geyser.GeyserImpl;
+import org.geysermc.geyser.api.item.custom.CustomItemData;
+import org.geysermc.geyser.api.item.custom.NonVanillaCustomItemData;
+import org.geysermc.geyser.item.GeyserCustomItemManager;
+import org.geysermc.geyser.item.GeyserCustomRenderOffsets;
+import org.geysermc.geyser.item.GeyserCustomMappingData;
+import org.geysermc.geyser.item.tools.ToolBreakSpeeds;
+import org.geysermc.geyser.registry.Registries;
+import org.geysermc.geyser.registry.type.ItemMapping;
+import org.geysermc.geyser.registry.type.ItemMappings;
+
+import java.util.*;
+import java.util.regex.Pattern;
+
+public class CustomItemRegistryPopulator {
+    private static IntSet protocolVersions = Registries.ITEMS.get().keySet();
+
+    public static GeyserCustomMappingData populateRegistry(String baseItem, CustomItemData customItemData, int customItems) {
+        String customItemName = GeyserCustomItemManager.CUSTOM_PREFIX + customItemData.name();
+
+        GeyserCustomMappingData returnData = new GeyserCustomMappingData();
+        int addNum = 0;
+
+        for (Int2ObjectMap.Entry<ItemMappings> entry : Registries.ITEMS.get().int2ObjectEntrySet()) {
+            addNum++;
+
+            int protocolVersion = entry.getIntKey();
+            ItemMappings itemMappings = entry.getValue();
+
+            ItemMapping javaItem = itemMappings.getMapping(baseItem);
+            if (javaItem == null) {
+                GeyserImpl.getInstance().getLogger().error("Could not find the java item to add custom model data to for " + baseItem + " in protocol version " + protocolVersion);
+                continue;
+            }
+
+            int nameExists = 0;
+            for (String mappingName : itemMappings.getCustomIdMappings().values()) {
+                if (Pattern.matches("^" + customItemName + "(_([0-9])+)?$", mappingName)) {
+                    nameExists++;
+                }
+            }
+            if (nameExists != 0) {
+                customItemName = customItemName + "_" + nameExists; //Since we update the name outside the for loop, this won't be run on each protocol version
+                GeyserImpl.getInstance().getLogger().warning("Custom item name '" + customItemData.name() + "' already exists, new item was registered as " + customItemName.substring(GeyserCustomItemManager.CUSTOM_PREFIX.length()) + "!");
```

Comment:
Do we want to support this usecase? I'd say that identifiers should be unique and that we should reject (or overwrite) them instead. And I don't think that people would (and should) expect them to be registered under a different name.

Reply:
Apparently, items adder does it. But I guess it's still down to what you map it as.